File: Python/sentimental-credit/credit.py
from cs50 import get_string


def Main():

    card_number = check_input()
    number_length = len(card_number)

    if checksum(card_number):
        check_processor(card_number, number_length)


# Checks input from user for numeric digits only and number length
def check_input():
    while (True):
        card_number = get_string("Type your credit card number: ")
        if card_number.isdigit():
            if len(card_number) >= 10:
                break
    return card_number


def checksum(card_number):
    # Creates an empty list to allocate card numbers
    numbers = []
    count = 0

    # Inserts numbers
    for c in card_number:
        numbers.insert(count, c)
        count += 1

    # Creates new list with reversed card number
    numbers = list(reversed(numbers))
    sum = 0

    # Iterates through every other number second-to-last
    for number in numbers[1::2]:
        digit = (int(number) * 2)
        if digit >= 10:
            sum += (digit % 10)
            digit = int(str(digit)[0:1:1])
            sum += digit
        else:
            sum += digit

    # Iterates through remaining numbers
    for number in numbers[::2]:
        digit = int(number)
        sum += digit

    # Checksum
    if sum % 10 == 0:
        return True
    else:
        invalid()


# Checks processor
def check_processor(card_number, number_length):
    # Checks for AMEX
    if number_length == 15:
        if int(card_number[:2]) == 34 or int(card_number[:2]) == 37:
            print("AMEX")
        else:
            invalid()

    # Checks for Mastercard and Visa
    elif number_length == 16:
        if int(card_number[:2]) >= 51 and int(card_number[:2]) <= 55:
            print("MASTERCARD")
        elif int(card_number[:1]) == 4:
            print("VISA")
        else:
            invalid()

    # Checks for Visa
    elif number_length == 13 and int(card_number[:1]) == 4:
        print("VISA")

    else:
        invalid()


# In case card checks as invalid throughout the program
def invalid():
    print("INVALID")
    return 1


Main()

File: 1-C/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

void checkProcessor(int length, int first, int second);

int main(void)
{

    long number = get_long("Type your Credit Card number: ");
    int number_length = floor(log10(number)) + 1;

    int digit[number_length];
    int sum = 0, multiplied_digit = 0;

    //Adds each number to a specific position of the array "digit".
    for (int i = 0; i < number_length; i++)
    {
        digit[i] = number % 10;
        number = number / 10;
    }

    //Manipulates every other digit starting second-to-last
    for (int i = 0; i < number_length; i++)
    {
        if (i % 2 == 0)
        {
            sum += digit[i];
        }
        else
        {
            multiplied_digit = digit[i] * 2;
            if (multiplied_digit >= 10)   //Adds the product's individual digits if greater than 9
            {
                sum += (multiplied_digit % 10);
                multiplied_digit = multiplied_digit / 10;
                sum += multiplied_digit;
            }
            else    //Adds the product's individual digits if smaller than 10
            {
                sum += multiplied_digit;
            }
        }
    }

    if (sum % 10 == 0)
    {
        checkProcessor(number_length, digit[(number_length - 1)], digit[(number_length - 2)]);
    }
    else
    {
        printf("INVALID\n");
    }
}

//Checks length and first digits then prints answer
void checkProcessor(int length, int first, int second)
{
    if (length == 15)
    {
        if (first == 3 && (second == 4 || second == 7))
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if (length == 13 || length == 16)
    {
        if (first == 5 && second > 0 && second < 6)
        {
            printf("MASTERCARD\n");
        }
        else if (first == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}